src/main/java/com/android/volley/RequestQueue.java

```diff
@@ -38,11 +38,29 @@
     // TODO: This should not be a generic class, because the request type can't be determined at
     // compile time, so all calls to onRequestFinished are unsafe. However, changing this would be
     // an API-breaking change. See also: https://github.com/google/volley/pull/109
+    @Deprecated // Use RequestEventListener instead.
     public interface RequestFinishedListener<T> {
         /** Called when a request has finished processing. */
         void onRequestFinished(Request<T> request);
     }
 
+    /** Request event types listeners will be notified about. */
+    public enum RequestEvent {
+        REQUEST_QUEUED,
+        REQUEST_CACHE_LOOKUP_STARTED,
+        REQUEST_CACHE_LOOKUP_FINISHED,
+        REQUEST_PROCESSING_STARTED,
+        REQUEST_PROCESSING_FINISHED,
+        REQUEST_FINISHED
+    }
+
+    /** Callback interface for request life cycle events. */
+    public interface RequestEventListener {
+        // Listener can be called from different threads but in synchronized way. Also it's
+        // guaranteed that the events will be sent in the order they have happened.
+        void onRequestEvent(Request request, RequestEvent event);
```

Comment:
Request<?> (otherwise this is a Java warning AFAIK)

Reply:
I actually think this doesn't make sense because all Requests must be passed to registered event listeners, even if they don't match the expected generic type (which is unknown at runtime). So while it compiles, it means there's an unsafe cast when passing the request to the event listener, and there's not actually any type safety. <?> accurately notes that this is a Request of unknown type, so we can't make any conclusions about it at compile time.

(See also the note above RequestFinishedListener)

Reply:
Ahh right, I forget it would be for all Requests :/

Reply:
Done